toolkit/tools/imagegen/diskutils/diskutils.go

```diff
@@ -217,6 +217,64 @@ func SetupLoopbackDevice(diskFilePath string) (devicePath string, err error) {
 	return
 }
 
+// BlockOnDiskIO waits until all outstanding operations against a disk complete.
+func BlockOnDiskIO(diskDevPath string) (err error) {
+	const (
+		majIdx            = 0
+		minIdx            = 1
+		outstandingOpsIdx = 11
+	)
+
+	logger.Log.Infof("Flushing all IO to disk for %s", diskDevPath)
+	shell.Execute("sync")
+
+	stdout, stderr, err := shell.Execute("lsblk", "--output", "MAJ:MIN", "--noheadings", "--nodeps", diskDevPath)
+	if err != nil {
+		logger.Log.Errorf("Failed to get disk major/minor ID (%s): %s ", err.Error(), stderr)
+		return
+	}
+	// lsblk includes whitespace both before an after output, trim it off.
+	diskIDs := strings.Split(strings.TrimSpace(stdout), ":")
+	if len(diskIDs) != 2 {
+		return fmt.Errorf("couldn't find disk IDs for %s (%s)", diskDevPath, stdout)
+	}
+	maj := diskIDs[majIdx]
+	min := diskIDs[minIdx]
+
+	logger.Log.Tracef("Searching /proc/diskstats for %s:%s", maj, min)
+	outstandingOps := ""
+	for busy := true; busy; busy = (outstandingOps != "0") {
```

Comment:
Might be worth adding a sleep somewhere in this loop to not overload the device with having to read `/proc/diskstats` all the time.